apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/MemoryCache.kt

```diff
@@ -0,0 +1,99 @@
+package com.apollographql.apollo.cache.normalized
+
+import com.apollographql.apollo.cache.ApolloCacheHeaders
+import com.apollographql.apollo.cache.CacheHeaders
+import com.apollographql.apollo.cache.normalized.internal.LruCache
+import com.apollographql.apollo.cache.normalized.internal.Platform
+import okio.internal.commonAsUtf8ToByteArray
+import kotlin.reflect.KClass
+
+class MemoryCache(
+    private val maxSizeBytes: Int,
+    private val expireAfterMillis: Long = -1,
+) : NormalizedCache() {
+  private val lruCache = LruCache<String, CacheEntry>(maxSize = maxSizeBytes) { key, cacheEntry ->
+    key.commonAsUtf8ToByteArray().size + (cacheEntry?.sizeInBytes ?: 0)
+  }
+
+  val size: Int
+    get() = lruCache.size()
+
+  override fun loadRecord(key: String, cacheHeaders: CacheHeaders): Record? {
+    val cachedEntry = lruCache[key]
+    return if (cachedEntry == null || cachedEntry.isExpired) {
+      if (cachedEntry != null) {
+        lruCache.remove(key)
+      }
+      nextCache?.loadRecord(key, cacheHeaders)
+    } else {
+      if (cacheHeaders.hasHeader(ApolloCacheHeaders.EVICT_AFTER_READ)) {
+        lruCache.remove(key)
+      }
+      cachedEntry.record
+    }
+  }
+
+  override fun clearAll() {
+    lruCache.clear()
+    nextCache?.clearAll()
+  }
+
+  override fun remove(cacheKey: CacheKey, cascade: Boolean): Boolean {
+    val cachedEntry = lruCache.remove(cacheKey.key)
+    if (cascade && cachedEntry != null) {
+      for (cacheReference in cachedEntry.record.referencedFields()) {
+        remove(CacheKey(cacheReference.key), true)
+      }
+    }
+
+    val removeFromNextCacheResult = nextCache?.remove(cacheKey, cascade) ?: false
+
+    return cachedEntry != null || removeFromNextCacheResult
+  }
+
+  override fun performMerge(apolloRecord: Record, oldRecord: Record?, cacheHeaders: CacheHeaders): Set<String> {
+    return if (oldRecord == null) {
+      lruCache[apolloRecord.key] = CacheEntry(
+          record = apolloRecord,
+          expireAfterMillis = expireAfterMillis
+      )
+      apolloRecord.keys()
+    } else {
+      oldRecord.mergeWith(apolloRecord).also {
+        //re-insert to trigger new weight calculation
+        lruCache[apolloRecord.key] = CacheEntry(
+            record = oldRecord,
+            expireAfterMillis = expireAfterMillis
+        )
+      }
+    }
+  }
+
+  @OptIn(ExperimentalStdlibApi::class)
+  override fun dump() = buildMap<KClass<*>, Map<String, Record>> {
```

Comment:
Same here, I think this could be handled in a more global way but we can definitely do this better.

Reply:
Hm, not sure if it's possible on higher lever as we need to get this map anyway from the low level cache impl

Reply:
Yea, I see it now, SQLite optimization require these extra APIs